### PR TITLE
Save and load  in NF4 / FP4 formats

### PR DESCRIPTION
**Pull request by poedator**
Purpose: enable saving and loading transformers models in 4bit formats.
Enables this PR in transformers: https://github.com/huggingface/transformers/pull/26037

addresses feature request #603 and other similar ones elsewhere.

tested with Bloom-560 and Llama-2-7b. 
tested quantization, saving, loading,  -- matched tensors and quant_states -- matched inference results.
tested both nf4 and fp4
test added

### CODE REVIEW

**Comment by poedator** [comment]
## on quant_state
While this PR does what is intended, it is important to make a flexible and extensible way to store quantized weights. This means that if the quant_state format should be refactored, it is best to do it now. Otherwise the future versions of `bnb` and `transformers` would have to support multitude of quant_state formats. I suggest making it a DataClass or NamedTuple or dict. With your blessing I can try this refactoring.

File: bitsandbytes/functional.py [hunk 1]
```diff
@@ -1472,23 +1566,22 @@ def gemv_4bit(
     out: Tensor = None,
     transposed_A=False,
     transposed_B=False,
-    state=None
```
Comment (TimDettmers) [inline bitsandbytes/functional.py:1475]: While quant_state is preferable, this would break code that used the different keyword. As such, to ensure backwards compatibility with the API, its better to leave the name as `state`. One could deprecate this, but I think its not worth it and we should just leave it as `state`

File: bitsandbytes/functional.py [hunk 2]
```diff
@@ -1798,7 +1891,7 @@ def igemmlt(A, B, SA, SB, out=None, Sout=None, dtype=torch.int32):
 
 def mm_dequant(
     A,
-    quant_state,
```
Comment (TimDettmers) [inline bitsandbytes/functional.py:1801]: Same as above. Althought, it is very unlikely that a 3rd party code uses this function, its better to leave the name as `state` to ensure backward compatibility of 3rd party code.

File: bitsandbytes/nn/modules.py [hunk 3]
```diff
@@ -178,22 +189,9 @@ def to(self, *args, **kwargs):
         if (device is not None and device.type == "cuda" and self.data.device.type == "cpu"):
             return self.cuda(device)
         else:
-            s = self.quant_state
-            if s is not None:
-                # make sure the quantization state is on the right device
-                s[0] = s[0].to(device)
-                if self.compress_statistics:
```
Comment (TimDettmers) [inline bitsandbytes/nn/modules.py:185]: Thank you for cleaning up this mess! :pray:

File: tests/test_linear4bit.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    sd = linear_q.state_dict()
+
+    # restoring from state_dict:
+
+    sd = linear_q.state_dict()
+    bias_data2 = sd.pop("bias", None)
+    weight_data2 = sd.pop("weight")
+
+    weight2 = bnb.nn.Params4bit.from_prequantized(quantized_stats=sd, data=weight_data2)
```
Comment (TimDettmers) [inline tests/test_linear4bit.py:50]: I like this API!

File: tests/test_linear4bit.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    linear_q2 = bnb.nn.Linear4bit(
+        linear.in_features,
+        linear.out_features,
+        bias=bias,
+        compute_dtype=compute_dtype,
+        compress_statistics=compress_statistics,
+        quant_type=quant_type,
+        device=device,
+    )
+    linear_q2.weight = weight2.to(device)
```
Comment (TimDettmers) [inline tests/test_linear4bit.py:61]: Doing it like this will create some overhead since the Linear4bit class will be initialized with random weights. Is there a good way to go around this? Not the most important fix, but if we can fix it, then its much faster to load NF4 models since weight initialization is usually on the CPU.

File: tests/test_linear4bit.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+from bitsandbytes import functional as F
+from bitsandbytes.nn.modules import Linear4bit
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="this test requires a GPU")
+@pytest.mark.parametrize(
+    "quant_type, compress_statistics, bias",
+    list(product(["nf4", "fp4"], [False, True], [False, True])),
+)
+def test_linear4_state_dict(quant_type, compress_statistics, bias):
```
Comment (TimDettmers) [inline tests/test_linear4bit.py:19]: I think this test looks good. I think another test might be (1) init, (2) copy, (3) quantize, (4) save, (4) del model, (5) load model, (6) compare against reference. This makes sure that we go through the full loop the user is going through. Internally, this should be the same as load/save the state_dict, but it's good to check everything is loaded correctly.

**Review by TimDettmers — COMMENTED** [review]
This looks very good. Thank you in particular for the quant_state changes. It was a thorn in my eyes and glad you cleaned this up. Some minor changes:
- restore a variable name `quant_state` to `state` in two functions, since these might be used in 3rd party code as keywords, and we do not want to change the API.
- another test that does the full save + load loop

File: tests/test_linear4bit.py (hunk 6 above)
Comment (poedator) [reply tests/test_linear4bit.py:19]: I added such test in [my PR to transformers](https://github.com/huggingface/transformers/pull/26037) and it passes together with this PR. I hesitate to bring it to this repo because it would require bringing most of that PR code to the test here. 
Once the transformers PR goes thru -- this would be easy to implement, and I will be happy to make a follow-up PR here.

File: bitsandbytes/functional.py (hunk 2 above)
Comment (poedator) [reply bitsandbytes/functional.py]: done

File: bitsandbytes/functional.py (hunk 1 above)
Comment (poedator) [reply bitsandbytes/functional.py]: done.
My idea was to name references to `MatmulLtState` class and to `QuantState` class differently.
Now reverted to the original signature in both `gemv_4bit()` (where it expects QuantState) and in `mm_dequant()` (where I was wrong - it was just some old format of quant_state).

**Comment by poedator** [comment]
To improve backwards compatibility, I added method QuantState.__get_item__() that imitates the old nested list format.

I also noticed that some functions in `functional.py` expect some other (perhaps older) format of quant_state:
- `nvidia_transform()` and `transform()` expect `[shape, from_order]`
- `mm_dequant()` expects `[shape, ...]`
- some others expect `[shape, code]`
I did not touch those functions, because they work OK as is, and some are only involved in test (deprecated?).
You may want to look at anything that calls `state` by index in `functional.py`.

File: tests/test_linear4bit.py (hunk 5 above)
Comment (poedator) [reply tests/test_linear4bit.py:61]: done. 
I edited the test to avoid unnecessary weights init. Now the layer is created on `'meta'` device and then populated with quantized weight.

**Comment by webpolis** [comment]
Thanks @poedator .. this is amazing. I am now using your transformers + bitsandbytes "save4" versions locally and it's working perfectly: I could save a fine-tuned 4 bits model with LoRa with the adapters merged with the base model and inference goes smoothly.

**Comment by poedator** [comment]
@webpolis,
Thank you for your interest in testing this branch; I'm pleased to see that it’s working well. I would like to mention a minor caveat: while I do not anticipate any modifications, the save format has not been officially finalized. Therefore, please be aware of potential compatibility issues when loading your current saves after this PR and the associated transformers PR are merged.

**Comment by webpolis** [comment]
> @webpolis,
> Thank you for your interest in testing this branch; I'm pleased to see that it’s working well. I would like to mention a minor caveat: while I do not anticipate any modifications, the save format has not been officially finalized. Therefore, please be aware of potential compatibility issues when loading your current saves after this PR and the associated transformers PR are merged.
> 

I understand and I'm aware. I'm monitoring releases here and there, so I hope your work gets merged soon. It is indeed few steps forward to democratizing AI.

**Comment by webpolis** [comment]
@poedator I do have a strange issue to share.

As explained above, I'm using your _bitsandbytes + transformers_' **save4** branches. I train a model using LoRa, loaded in 4 bits using the following settings:

```python
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
    # llm_int8_enable_fp32_cpu_offload=True
)

model = AutoModelForCausalLM.from_pretrained(
    model_name_or_path,
    device_map=device_map,
    torch_dtype=torch.float16,
    quantization_config=bnb_config,
    use_cache=False,
)

peft_config = LoraConfig(
    r=16 # also tried 8,
    lora_alpha=32 # also tried 16,
    target_modules=['k_proj', 'out_proj', 'q_proj', 'v_proj', 'fc_in', 'fc_out'],
    lora_dropout=0.1,
    bias="none",
    task_type=TaskType.CAUSAL_LM,
    inference_mode=False
)

model.config.pretraining_tp = 1
model.gradient_checkpointing_enable()

model = prepare_model_for_kbit_training(model)
model = get_peft_model(model, peft_config)

training_args = TrainingArguments(
    per_device_train_batch_size=BATCHES,
    gradient_accumulation_steps=1,
    max_steps=MAX_STEPS,
    learning_rate=LR,
    lr_scheduler_type='linear',
    warmup_steps=WARMUP_STEPS,
    warmup_ratio=0.1,
    fp16=True,
    gradient_checkpointing=True,
    logging_steps=10,
    output_dir="out",
    optim="adamw_hf",
    save_total_limit=3,
    save_steps=100,
    # deepspeed=ds_config,
    report_to='none'
)
trainer = Trainer(
    model=model,
    train_dataset=train_ds,
    args=training_args,
    data_collator=partial(collate_fn)
)

trainer.train()
```

After training, I load the checkpoint:

```python
peft_model = PeftModel.from_pretrained(model, CHECKPOINT)
```

And do some inference:

```python
BEAMS = 1
streamer = TextStreamer(tokenizer)
inputs = tokenizer(prompt, return_tensors="pt")
input_ids = inputs["input_ids"].to(model.device)
attention_mask = inputs["attention_mask"].to(model.device)
generation_config = GenerationConfig(
    temperature=0.1,
    top_p=0.8,
    top_k=40,
    # penalty_alpha=0.6,
    # num_beams=BEAMS,
    repetition_penalty=1.5,
    do_sample=True,
    eos_token_id=tokenizer.eos_token_id,
    pad_token=tokenizer.pad_token_id,
    max_length=CTX_LEN,
    use_cache=True
)

print(f'{len(input_ids.squeeze())} tokens in prompt\n\n')

with torch.no_grad():
    generation_output = model.generate(
        input_ids=input_ids,
        pad_token_id=tokenizer.eos_token_id,
        attention_mask=attention_mask,
        generation_config=generation_config,
        return_dict_in_generate=True,
        output_scores=True,
        early_stopping=BEAMS > 1,
        streamer=streamer
    )
```

Things go great as of now, since I added an extra token while training and the response match the intention of such token when I inject it into the prompt. However, once I merge, save, restart and use the same token, it doesn't go further than a few letters or emptiness like:

```python
<|vulgarg|>oas<|endoftext|>
```

In summary: things go great by loading the LoRa checkpoints separately. But once I load the full saved merged model, it doesn't produce good data with the newly added tokens. If I remove the token, the model works perfect in both cases.

**Comment by webpolis** [comment]
To add up to above, I'm using a GPT-J model.

**Comment by webpolis** [comment]
I updated https://github.com/TimDettmers/bitsandbytes/pull/753#issuecomment-1740029533

**Comment by poedator** [comment]
> I updated [#753 (comment)](https://github.com/TimDettmers/bitsandbytes/pull/753#issuecomment-1740029533)

Hi, @webpolis ,
thank you for testing this PR. Could you check whether your example works OK with 8bit quantization? Please share complete code / notebook example for testing on my side.

File: tests/test_linear4bit.py (hunk 4 above)
Comment (poedator) [reply tests/test_linear4bit.py]: I liked it too, but by now replaced it with more conventionally-looking `from_dict` which accepts relevant part of `state_dict`

**Comment by poedator** [comment]
Just made couple of commits:
- using `state_dict` as way to load/save Linear4bit layer. All magic happens inside the layer methods so that from outside it could be loaded using regular ways
- renamed `code` component of saved quant_state into `quant_map`. It may be a good idea to rename the relevant variables too - `code` is not a very intuitive name.

**Comment by vidhyat98** [comment]
Eagerly looking for this PR to be merged!

**Comment by TimDettmers** [comment]
Looks all great, thank you!

**Comment by RonanKMcGovern** [comment]
@poedator many thanks.

- Does this mean that doing merge and unload is a small step away? for NF4?
- If a model is saved to hub in NF4, can I load it in bf16 (obviously with lost precision)?

**Comment by poedator** [comment]
> @poedator many thanks.
> 
> * Does this mean that doing merge and unload is a small step away? for NF4?
> * If a model is saved to hub in NF4, can I load it in bf16 (obviously with lost precision)?

Hi, @RonanKMcGovern ,

1) Hopefully it is a matter of days. Testing the transformers integration now - watch this PR: https://github.com/huggingface/transformers/pull/26037

2) the idea is to give the user after loading exactly the same model  as could be obtained from `.from_pretrained(..., load_in_4_bit=True)` (or with similar quantization config). The bnb_4bit_compute_dtype is set at the time of initial quantization. Presumeably it can be switched later. try `Linear4bit.set_compute_type()`.

**Comment by RonanKMcGovern** [comment]
@poedator I've being trying to push to hub in 4 bit but am failing.

## Replication
```
# Required when quantizing models/data that are gated on HuggingFace, and required for pushing models to HuggingFace.
!pip install huggingface_hub -q -U
from huggingface_hub import notebook_login

notebook_login()

!python -m pip install --upgrade pip
!pip install -U -q transformers
!pip install -q -U bitsandbytes
!pip install -q -U accelerate
!pip install -q -U scipy

model_id = "mistralai/Mistral-7B-Instruct-v0.2"

from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, AutoConfig
import torch

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config, #for loading in 4 bit.
    device_map='auto',
    # trust_remote_code=False,
    torch_dtype=torch.bfloat16,
    cache_dir='')

model.push_to_hub(
    new_repo,
    token=True,
    max_shard_size="10GB",
    use_safetensors=True)
```

Error:
```
Cell In[37], line 1
----> 1 model.push_to_hub(
      2     new_repo,
      3     revision=revision,
      4     token=True,
      5     max_shard_size="10GB",
      6     use_safetensors=True)

File /usr/local/lib/python3.10/dist-packages/transformers/utils/hub.py:871, in push_to_hub(self, repo_id, use_temp_dir, commit_message, private, token, max_shard_size, create_pr, safe_serialization, revision, commit_description, **deprecated_kwargs)
    864 # Save all files.
    865 self.save_pretrained(work_dir, max_shard_size=max_shard_size, safe_serialization=safe_serialization)
    867 return self._upload_modified_files(
    868     work_dir,
    869     repo_id,
    870     files_timestamps,
--> 871     commit_message=commit_message,
    872     token=token,
    873     create_pr=create_pr,
    874     revision=revision,
    875     commit_description=commit_description,
    876 )

File /usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py:2206, in PreTrainedModel.save_pretrained(self, save_directory, is_main_process, state_dict, save_function, push_to_hub, max_shard_size, safe_serialization, variant, token, save_peft_format, **kwargs)
   2204 # If the model has adapters attached, you can save the adapters
   2205 if getattr(self, "is_loaded_in_4bit", False) and not _hf_peft_config_loaded:
-> 2206     raise NotImplementedError(
   2207         "You are calling `save_pretrained` on a 4-bit converted model. This is currently not supported"
   2208     )
   2210 if getattr(self, "_awq_is_fused", False):
   2211     raise ValueError("You cannot save an AWQ model that uses fused modules!")

NotImplementedError: You are calling `save_pretrained` on a 4-bit converted model. This is currently not supported
```

## Other things I tried
- Llama 2 model, Mixtral Model
- Using load_in_4bit instead of the bnb config
- Updating to dev versions of bitsandbytes and transformers

## Transformers version
- `transformers` version: 4.36.2
- Platform: Linux-5.4.0-153-generic-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.2
- Safetensors version: 0.4.1
- Accelerate version: 0.25.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.0+cu118 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: A6000
- Using distributed or parallel set-up in script?: Single GPU

Bitsandbytes version is bitsandbytes-0.42.0

**Comment by poedator** [comment]
Hi, @RonanKMcGovern! Thank you for testing.

the bnb 4bit serialization feature is not part of transformers==4.36.2.
You may want to install from the latest github code:
`pip install git+https://github.com/huggingface/transformers.git`
or wait to 4.37 (hopefully).

**Comment by RonanKMcGovern** [comment]
> the bnb 4bit serialization feature is not part of transformers==4.36.2.
> You may want to install from the latest github code:
> `pip install git+https://github.com/huggingface/transformers.git`
> or wait to 4.37 (hopefully).

Thanks! This is working well! Hopefully yeah it gets into 4.37

I'm also hoping TGI will allow these serialised models to be directly loaded, would be nice. [Issue here](https://github.com/huggingface/text-generation-inference/issues/1410)

**Comment by chynphh** [comment]
Hi, @poedator 

I have a problem with quantifying [Qwen-VL](https://github.com/QwenLM/Qwen-VL/). There is no problem when saving the model, but an error is reported when reloading it.

My code is as follows:
```python
import os, sys, json
import time
import sys
import torch
from transformers import BitsAndBytesConfig
from transformers import AutoModelForCausalLM, AutoTokenizer


tokenizer = AutoTokenizer.from_pretrained(model_dir, trust_remote_code=True)

bnb_model_from_pretrained_args = {}
bnb_model_from_pretrained_args.update(dict(
    load_in_4bit=True,
    load_in_8bit=False,
    quantization_config=BitsAndBytesConfig(
        load_in_4bit=True,
        load_in_8bit=False,
        llm_int8_threshold=6.0,
        llm_int8_has_fp16_weight=False,
        bnb_4bit_compute_dtype=torch.bfloat16,
        bnb_4bit_use_double_quant=True,
        bnb_4bit_quant_type='nf4' # {'fp4', 'nf4'}
    )
))

model = AutoModelForCausalLM.from_pretrained(model_dir, device_map="cuda", trust_remote_code=True, **bnb_model_from_pretrained_args)

model.save_pretrained(my_local_quant_dir)
model = AutoModelForCausalLM.from_pretrained(my_local_quant_dir, load_in_4bit=True, trust_remote_code=True)
```


The error reported is as follows:
```

Some weights of the model checkpoint a /mnt/workspace/models/test-qw2 were not used when initializing QWenLMHeadModel: ['transformer.visual.transformer.resblocks.18.mlp.c_proj.weight.absmax', 'transformer.visual.transformer.resblocks.18.mlp.c_proj.weight.nested_absmax', 'transformer.visual.transformer.resblocks.18.mlp.c_proj.weight.nested_quant_map', 'transformer.visual.transformer.resblocks.18.mlp.c_proj.weight.quant_map', 'transformer.visual.transformer.resblocks.18.mlp.c_proj.weight.quant_state.bitsandbytes__nf4']
- This IS expected if you are initializing QWenLMHeadModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing QWenLMHeadModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Traceback (most recent call last):
  File "/mnt/workspace/x-workspace/project/v0/test/merge_and_quant.py", line 254, in <module>
    model = AutoModelForCausalLM.from_pretrained(my_local_quant_dir, load_in_4bit=True, trust_remote_code=True)
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/transformers/models/auto/auto_factory.py", line 561, in from_pretrained
    return model_class.from_pretrained(
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/transformers/modeling_utils.py", line 3917, in from_pretrained
    dispatch_model(model, **device_map_kwargs)
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/accelerate/big_modeling.py", line 399, in dispatch_model
    attach_align_device_hook_on_blocks(
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/accelerate/hooks.py", line 517, in attach_align_device_hook_on_blocks
    add_hook_to_module(module, hook)
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/accelerate/hooks.py", line 156, in add_hook_to_module
    module = hook.init_hook(module)
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/accelerate/hooks.py", line 254, in init_hook
    set_module_tensor_to_device(module, name, self.execution_device)
  File "/mnt/workspace/miniconda3/envs/py39mm/lib/python3.9/site-packages/accelerate/utils/modeling.py", line 306, in set_module_tensor_to_device
    raise ValueError(f"{tensor_name} is on the meta device, we need a `value` to put in on {device}.")
ValueError: weight is on the meta device, we need a `value` to put in on 0.
```

transformers 4.37.2
bitsandbytes 0.42.0

**Comment by poedator** [comment]
Hi, @chynphh 
This may happen if you used earlier version of the code when saving the model. 
please check if you used same versions when loading and saving the model.

**Comment by ahyunsoo3** [comment]
After training 4bit with LoRA, merging with original base and saving it, there is no error. Thanks great work @poedator.
